File: library/src/main/java/wannabe/grid/Grid.java
package wannabe.grid;

import java.util.Iterator;

import wannabe.Voxel;

/**
 * A collection of {@link Voxel}s. Voxels do not move within to the grid, although the grid's
 * set of voxels may change (see {@link MutableGrid}) or the translation of the grid itself
 * may change.
 */
public interface Grid extends Iterable<Voxel> {
  /**
   * Returns an iterator over all {@link Voxel}s in this grid. Note that optimized implementations
   * may return the same {@link Voxel} reference with every call to {@link Iterator#next()} so it is
   * not recommended that references are kept after use.
   */
  @Override Iterator<Voxel> iterator();

  /** Number of Voxels this grid contains. */
  int size();

  /**
   * Returns {@code true} if anything about this grid has changed (such as voxels or translation)
   * since the last time {@link #iterator()} was called.
   */
  boolean isDirty();

  AllNeighbors neighbors(Voxel voxel);

//  don't remember what I was doing here:
//  // Painting methods.
//  Voxel at(Translation translation);
}

File: library/src/main/java/wannabe/grid/Grid.kt
package wannabe.grid

import wannabe.Voxel

/**
 * A collection of [Voxel]s. Voxels do not move within to the grid, although the grid's
 * set of voxels may change (see [MutableGrid]) or the translation of the grid itself
 * may change.
 */
interface Grid : Iterable<Voxel> {
  /** Returns an iterator over all [Voxel]s in this grid. */
  override fun iterator(): Iterator<Voxel>

  /** Number of Voxels this grid contains.  */
  val size: Int

  /**
   * Returns `true` if anything about this grid has changed (such as voxels or translation)
   * since the last time [.iterator] was called.
   */
  val isDirty: Boolean

  /**
   * Gets the neighbor information for the specified [Voxel]. Implementations may reuse the
   * returned instance, so it's only valid to use until the next [neighbors()] call
   */
  fun neighbors(voxel: Voxel): AllNeighbors
}
